go/coordinator/internal/grpccoordinator/collection_service.go

```diff
@@ -0,0 +1,292 @@
+package grpccoordinator
+
+import (
+	"context"
+
+	"github.com/chroma/chroma-coordinator/internal/common"
+	"github.com/chroma/chroma-coordinator/internal/model"
+	"github.com/chroma/chroma-coordinator/internal/proto/coordinatorpb"
+	"github.com/chroma/chroma-coordinator/internal/types"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const errorCode = 500
+const successCode = 200
+const success = "ok"
+
+func (s *Server) ResetState(context.Context, *emptypb.Empty) (*coordinatorpb.ChromaResponse, error) {
+	res := &coordinatorpb.ChromaResponse{}
+	err := s.coordinator.ResetState(context.Background())
+	if err != nil {
+		res.Status = failResponseWithError(err, errorCode)
+		return res, err
+	}
+	setResponseStatus(successCode)
+	return res, nil
+}
+
+func (s *Server) CreateCollection(ctx context.Context, req *coordinatorpb.CreateCollectionRequest) (*coordinatorpb.CreateCollectionResponse, error) {
+	getOrCreate := req.GetGetOrCreate()
+	if getOrCreate {
+		return s.getOrCreateCollection(ctx, req)
+	} else {
+		return s.createCollection(ctx, req)
+	}
+}
+
+// Cases for get_or_create
+
+// Case 0
+// new_metadata is none, coll is an existing collection
+// get_or_create should return the existing collection with existing metadata
+// Essentially - an update with none is a no-op
+
+// Case 1
+// new_metadata is none, coll is a new collection
+// get_or_create should create a new collection with the metadata of None
+
+// Case 2
+// new_metadata is not none, coll is an existing collection
+// get_or_create should return the existing collection with updated metadata
+
+// Case 3
+// new_metadata is not none, coll is a new collection
+// get_or_create should create a new collection with the new metadata, ignoring
+// the metdata of in the input coll.
+
+// The fact that we ignore the metadata of the generated collections is a
+// bit weird, but it is the easiest way to excercise all cases
+func (s *Server) getOrCreateCollection(ctx context.Context, req *coordinatorpb.CreateCollectionRequest) (*coordinatorpb.CreateCollectionResponse, error) {
+	res := &coordinatorpb.CreateCollectionResponse{}
```

Comment:
I think we should push the get_or_create down into the lower levels so we can get transactional writes. Can you open an issue?